### PR TITLE
fix(src): handle new form of list item as paragraph - I192

### PR DESCRIPTION
**Pull request by irmerk**
# Issue #192
Adjust the handling of lists with the new `paragraph` support in `markdown-transform v0.8`

### Changes
- Keep focus when clicking a block in the toolbar
- Handle block types as constants
- Handle breakout of empty list item
- Modularize and refactor code
  - Block action methods
  - `onClickBlock` handler
  - Handle lists `onEnter` in the plugin

### Flags
- Work still needed for this PR
  - ~Persisting issue around `block_quote` to `list_item`, requires further work~
  - ~Formatting Toolbar buttons don't yet recognize the selection being in a list~
- Future Issues
  - Work around `SHIFT + ENTER` handling a `linebreak` / new `paragraph` in a `list_item` is needed

### Related Issues
- N/A

### CODE REVIEW

**Comment by irmerk** [comment]
### [Netlify deployment](https://deploy-preview-194--accordproject-markdown-editor.netlify.com/examples/)

File: src/plugins/list.js [hunk 1]
```diff
@@ -11,7 +13,35 @@ function List() {
...
+      return false;
+    }
+
+    // Hitting enter on a non-empty list item will add a new list_item
+    if (isSelectionList(value) && startBlock.text.length !== 0) {
+      editor.withoutNormalizing(() => {
+        event.preventDefault();
+        editor.splitBlock().unwrapBlock(CONST.LIST_ITEM).wrapBlock(CONST.LIST_ITEM);
+      });
+      return true;
```
Comment (mttrbrts) [inline src/plugins/list.js]: What is the effect of this? Shouldn't it return `next()` instead?

Comment (irmerk) [reply src/plugins/list.js]: I think `next()` is resulting in there being a `paragraph` in the `list_item` and a new `paragraph` below, as if a sibling of the `list_item`.

File: src/schema.js [hunk 2]
```diff
@@ -89,8 +89,6 @@ const schema = {
       nodes: [
         {
           match: [
-            { object: 'text' },
-            { type: 'link' },
```
Comment (jeromesimeon) [inline src/schema.js]: Shouldn't this include lists as well?

File: src/plugins/list.js [hunk 3]
```diff
@@ -11,7 +13,35 @@ function List() {
...
+    event.preventDefault();
+
+    // Hitting enter on a blank list item will break out of the enclosing list
+    if (isSelectionList(value) && startBlock.text.length === 0) {
+      editor.withoutNormalizing(() => {
+        event.preventDefault();
+        editor
+          .setBlocks(CONST.PARAGRAPH)
+          .unwrapBlock(CONST.LIST_ITEM)
+          .unwrapBlock(CONST.OL_LIST)
```
Comment (jeromesimeon) [inline src/plugins/list.js]: I do not understand the double unwrapping of first `OL_LIST` and then `UL_LIST`. Shouldn't this unwrap only one? (i.e., the first enclosing list?)

File: src/plugins/list.js (hunk 1 above)
Comment (jeromesimeon) [reply src/plugins/list.js]: I'm confused about this as well. What should be the type returned by `onEnter` ?

File: src/schema.js (hunk 2 above)
Comment (irmerk) [reply src/schema.js]: I believe so, but I think that should be left for another issue when we add support for nested lists.

File: src/plugins/list.js (hunk 1 above)
Comment (irmerk) [reply src/plugins/list.js]: I'm unsure if it should be returning a type. @dselman you may have some insight?

File: src/plugins/list.js (hunk 3 above)
Comment (irmerk) [reply src/plugins/list.js]: Good point. I think Slate was handling this as one unwrap to whichever list it was in. I refactored to unwrap once for whichever list it was.

File: src/plugins/list.js (hunk 1 above)
Comment (DianaLease) [reply src/plugins/list.js]: What is `.unwrapBlock(CONST.LIST_ITEM).wrapBlock(CONST.LIST_ITEM)` doing on line 37? Why would we unwrap a list item just to wrap it again?

File: src/plugins/list.js [hunk 4]
```diff
@@ -11,7 +13,34 @@ function List() {
...
+          .unwrapBlock(currentList(value).type);
+      });
+      return false;
+    }
+
+    // Hitting enter on a non-empty list item will add a new list_item
+    if (isSelectionInput(value, CONST.LIST_ITEM) && startBlock.text.length !== 0) {
+      editor.withoutNormalizing(() => {
+        event.preventDefault();
+        editor.splitBlock().unwrapBlock(CONST.LIST_ITEM).wrapBlock(CONST.LIST_ITEM);
```
Comment (DianaLease) [inline src/plugins/list.js]: What is .unwrapBlock(CONST.LIST_ITEM).wrapBlock(CONST.LIST_ITEM) doing on line 37? Why would we unwrap a list item just to wrap it again?

Comment (irmerk) [reply src/plugins/list.js]: I made this diagram that hopefully helps explain the process.

Image from iOS (1)

Comment (DianaLease) [reply src/plugins/list.js]: Thanks for explaining this. Makes sense now!

**Pull request merged**